### PR TITLE
Support containerized kubelet in CI

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:

  * Adjust hyperkube Makefile to allow HYPERKUBE_BIN to be
    specified from outside

  * new DOCKER_ROOT variable to specify --root-dir and associated
    volume for kubelet to know where docker directory lives

  * Save the docker logs when we kill the kubelet container

  * Allow customized image for dockerized kubelet and build a Default
    image if one is not specified

  We should allow a custom build using script we have:
  `VERSION="latest" REGISTRY="k8s.gcr.io" hack/dev-push-hyperkube.sh`

  to be used quickly using DOCKERIZE_KUBELET. To do this we remove the
  hard coded `k8s.gcr.io/kubelet` image and introduce a new environment
  variable `KUBELET_IMAGE`.

  Note that we are switching to hyperkube from kubelet as we have a quick
  and easy script to build hyperkube image and load it into local docker
  daemon. This reduces bad hacks like `docker tag
  k8s.gcr.io/hyperkube-amd64:<tag> k8s.gcr.io/kubelet:latest` being used
  by folks today (see 62057)

  * Better cope with PID 0 when running "docker inspect" to look for
    the process id for kubelet container

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -30,6 +30,7 @@ RUNTIME_CONFIG=${RUNTIME_CONFIG:-""}
 KUBELET_AUTHORIZATION_WEBHOOK=${KUBELET_AUTHORIZATION_WEBHOOK:-""}
 KUBELET_AUTHENTICATION_WEBHOOK=${KUBELET_AUTHENTICATION_WEBHOOK:-""}
 POD_MANIFEST_PATH=${POD_MANIFEST_PATH:-"/var/run/kubernetes/static-pods"}
+KUBELET_IMAGE={KUBELET_IMAGE:"k8s.gcr.io/hyperkube-amd64:latest"}
```
Comment (sttts) [inline hack/local-up-cluster.sh]: `:-`

Comment (dims) [reply hack/local-up-cluster.sh]: Dang! fixing.

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -30,6 +30,7 @@ RUNTIME_CONFIG=${RUNTIME_CONFIG:-""}
 KUBELET_AUTHORIZATION_WEBHOOK=${KUBELET_AUTHORIZATION_WEBHOOK:-""}
 KUBELET_AUTHENTICATION_WEBHOOK=${KUBELET_AUTHENTICATION_WEBHOOK:-""}
 POD_MANIFEST_PATH=${POD_MANIFEST_PATH:-"/var/run/kubernetes/static-pods"}
+KUBELET_IMAGE={KUBELET_IMAGE:"k8s.gcr.io/hyperkube-amd64:-latest"}
```
Comment (cblecker) [inline hack/local-up-cluster.sh]: This should be `KUBELET_IMAGE={KUBELET_IMAGE:-"k8s.gcr.io/hyperkube-amd64:-latest"}`

Comment (dims) [reply hack/local-up-cluster.sh]: i thought i fixed this earlier .. looks like i forgot to push it up.

Comment (dims) [reply hack/local-up-cluster.sh]: Done. thanks @cblecker

File: hack/local-up-cluster.sh [hunk 3]
```diff
@@ -30,6 +30,7 @@ RUNTIME_CONFIG=${RUNTIME_CONFIG:-""}
 KUBELET_AUTHORIZATION_WEBHOOK=${KUBELET_AUTHORIZATION_WEBHOOK:-""}
 KUBELET_AUTHENTICATION_WEBHOOK=${KUBELET_AUTHENTICATION_WEBHOOK:-""}
 POD_MANIFEST_PATH=${POD_MANIFEST_PATH:-"/var/run/kubernetes/static-pods"}
+KUBELET_IMAGE={KUBELET_IMAGE:-"k8s.gcr.io/hyperkube-amd64:latest"}
```
Comment (cblecker) [inline hack/local-up-cluster.sh]: The other thing is.. are you looking for an image that exists? The `latest` tag doesn't exist for the production image.

Comment (dims) [reply hack/local-up-cluster.sh]: Guess it would be better to leave ":latest" off

**Comment by cofyc** [comment]
How about don't give `KUBELET_IMAGE` default value? When it is unset, build kubelet image automatically. Like `GO_OUT`, `local-up-cluster.sh` builds binaries if `GO_OUT` is unset.

Some use cases:

```
DOCKERIZE_KUBELET=y
./hack/local-cluster-up.sh
```

builds `k8s.gcr.io/hyperkube-amd64:latest` image and uses it to run kubelet in container

```
DOCKERIZE_KUBELET=y
KUBELET_IMAGE=k8s.gcr.io/hyperkube-amd64:latest
./hack/local-cluster-up.sh
```

reuses previous built image

```
DOCKERIZE_KUBELET=y
KUBELET_IMAGE=k8s.gcr.io/hyperkube-amd64:v1.10.2-beta.0
./hack/local-cluster-up.sh -o ./_output/dockerized/bin/linux/amd64
```

uses `k8s.gcr.io/hyperkube-amd64:v1.10.2-beta.0` to run kubelet in container and uses previous built binaries to start apiserver, scheduler, etc.

**Comment by dims** [comment]
@cofyc +1 feel free to propose a PR with your idea.

**Comment by dims** [comment]
/hold

i am going to try adding the feature @cofyc mentioned

**Comment by cofyc** [comment]
@dims Great! Probably we can also add a flag to specify image, like what `-o` flag does.

cc @jsafrane

**Comment by dims** [comment]
@cofyc this is ready, please take a look.

**Comment by dims** [comment]
@mikedanese @ixdy can you please check/approve the changes to `cluster/images/hyperkube/Makefile`?

**Review by spiffxp — APPROVED** [review]
/lgtm

**Pull request merged**